### PR TITLE
Add VPC data source

### PR DESCRIPTION
**Pull request by ksamoray**

### CODE REVIEW

File: nsxt/data_source_policy_vpc_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+					if err := testAccDataSourceNsxtPolicyVPCCreate(name); err != nil {
+						t.Error(err)
+					}
+				},
+				Config: testAccNsxtPolicyVPCReadTemplate(name),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(testResourceName, "display_name", name),
+					resource.TestCheckResourceAttr(testResourceName, "description", name),
+					resource.TestCheckResourceAttrSet(testResourceName, "path"),
+					//resource.TestCheckResourceAttrSet(testResourceName, "default_gateway_path"),
```
Comment (salv-orlando) [inline nsxt/data_source_policy_vpc_test.go:43]: Why is this attribute commented? Is it giving you troubles?

File: nsxt/data_source_nsxt_policy_vpc.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		objGet, err := client.Get(defaultOrgID, ctx.ProjectID, objID)
+		if err != nil {
+			return handleDataSourceReadError(d, "VPC", objID, err)
+		}
+		obj = objGet
+	} else if objName == "" {
+		return fmt.Errorf("error obtaining VPC ID or name during read")
+	} else {
+		// Get by full name/prefix
+		objList, err := client.List(defaultOrgID, ctx.ProjectID, nil, nil, nil, nil, nil, nil)
```
[the remark this answers is not in the document]
Comment (ksamoray) [reply nsxt/data_source_nsxt_policy_vpc.go:55]: I've tried to implement the project datasource with search API at the time, which didn't work. So I didn't even try with this one - maybe it's worthwhile trying.

[the remark this answers is not in the document]
Comment (ksamoray) [reply nsxt/data_source_nsxt_policy_vpc.go:55]: client.List() call (L55) takes a project parameter, so uniqueness is only checked at the scope of these query results.

File: nsxt/data_source_nsxt_policy_vpc.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		for _, objInList := range objList.Results {
+			if strings.HasPrefix(*objInList.DisplayName, objName) {
+				prefixMatch = append(prefixMatch, objInList)
+			}
+			if *objInList.DisplayName == objName {
+				perfectMatch = append(perfectMatch, objInList)
+			}
+		}
+		if len(perfectMatch) > 0 {
+			if len(perfectMatch) > 1 {
```
[the remark this answers is not in the document]
Comment (ksamoray) [reply nsxt/data_source_nsxt_policy_vpc.go:71]: client.List() call (L55) takes a project parameter, so uniqueness is only checked at the scope of these query results.

File: nsxt/data_source_policy_vpc_test.go (hunk 1 above)
Comment (ksamoray) [reply nsxt/data_source_policy_vpc_test.go:43]: I didn't set the DGW path parameter in Create() (L84), I intended to remove this as although I could gather the T0 id, I had to create a path from it as I do with IP block. It's quite ugly... If you think that there's some importance to this, I can add it.
TBH I'd rather rewrite this once we have a resource for VPC.

File: nsxt/data_source_nsxt_policy_vpc.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+func dataSourceNsxtPolicyVPC() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceNsxtPolicyVPCRead,
+		Schema: map[string]*schema.Schema{
+			"id":                   getDataSourceIDSchema(),
+			"display_name":         getDataSourceDisplayNameSchema(),
+			"description":          getDataSourceDescriptionSchema(),
+			"path":                 getPathSchema(),
+			"context":              getContextSchema(),
```
Comment (annakhm) [inline nsxt/data_source_nsxt_policy_vpc.go:23]: Should project be required here?

File: nsxt/data_source_nsxt_policy_vpc.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (annakhm) [reply nsxt/data_source_nsxt_policy_vpc.go:55]: Should this support pagination?

File: nsxt/data_source_nsxt_policy_vpc.go (hunk 4 above)
Comment (ksamoray) [reply nsxt/data_source_nsxt_policy_vpc.go:23]: Indeed

File: nsxt/data_source_nsxt_policy_vpc.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ksamoray) [reply nsxt/data_source_nsxt_policy_vpc.go]: Replaced logic with search API

File: website/docs/d/policy_vpc.html.markdown [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+In addition to arguments listed above, the following attributes are exported:
+
+* `description` - The description of the resource.
+* `path` - The NSX path of the policy resource.
+* `short_id` - Defaults to id if id is less than equal to 8 characters or defaults to random generated id if not set.
+* `site_info` - Information related to sites applicable for given Project.
+    * `edge_cluster_paths` - The edge cluster on which the networking elements for the Org will be created.
+    * `site_path` - This represents the path of the site which is managed by Global Manager. For the local manager, if set, this needs to point to 'default'.
+* `default_gateway_path` - The tier 0 has to be pre-created before Project is created.
```
Comment (annakhm) [inline website/docs/d/policy_vpc.html.markdown:45]: This description does not seem relevant to data source

File: website/docs/d/policy_vpc.html.markdown [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+## Attributes Reference
+
+In addition to arguments listed above, the following attributes are exported:
+
+* `description` - The description of the resource.
+* `path` - The NSX path of the policy resource.
+* `short_id` - Defaults to id if id is less than equal to 8 characters or defaults to random generated id if not set.
+* `site_info` - Information related to sites applicable for given Project.
+    * `edge_cluster_paths` - The edge cluster on which the networking elements for the Org will be created.
```
Comment (annakhm) [inline website/docs/d/policy_vpc.html.markdown:43]: `Org` => `VPC`?

File: website/docs/d/policy_vpc.html.markdown [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    * `site_path` - This represents the path of the site which is managed by Global Manager. For the local manager, if set, this needs to point to 'default'.
```
Comment (annakhm) [inline website/docs/d/policy_vpc.html.markdown:44]: I'm not sure this is relevant for `VPC`

File: website/docs/d/policy_vpc.html.markdown [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  }
+  display_name = "vpc1"
+}
+```
+
+## Argument Reference
+
+* `id` - (Optional) The ID of VPC to retrieve. If ID is specified, no additional argument should be configured.
+* `display_name` - (Optional) The Display Name prefix of the VPC to retrieve.
+* `context` - (Optional) The context which the object belongs to
```
Comment (annakhm) [inline website/docs/d/policy_vpc.html.markdown:32]: This should be `Required`

File: nsxt/data_source_nsxt_policy_vpc.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+			"id":                   getDataSourceIDSchema(),
+			"display_name":         getDataSourceDisplayNameSchema(),
+			"description":          getDataSourceDescriptionSchema(),
+			"path":                 getPathSchema(),
+			"context":              getContextSchema(true, false),
+			"site_info":            getSiteInfoSchema(),
+			"default_gateway_path": getPathSchema(),
+			"short_id": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (annakhm) [inline nsxt/data_source_nsxt_policy_vpc.go:25]: According to the code this is fully computed, in which case `Optional: true` should be removed, so that terraform does not allow user to specify it

File: website/docs/d/policy_vpc.html.markdown (hunk 7 above)
Comment (ksamoray) [reply website/docs/d/policy_vpc.html.markdown:44]: I'm not sure either - I reused the SiteInfo from Project as these two use the same schema. Should I remove site_path entirely from this DS?

File: nsxt/data_source_nsxt_policy_vpc.go (hunk 9 above)
Comment (ksamoray) [reply nsxt/data_source_nsxt_policy_vpc.go:25]: This value is assigned by NSX if not set by user, and therefore has to be computed. It can also be set by the user (it should be unique - NSX will enforce that).
The UI doesn't allow updating it (on create there's the text "This field can be set at VPC creation time and cannot be edited later. Max 8 characters allowed.").
BTW there's an open issue for this in the Project.

File: website/docs/d/policy_vpc.html.markdown (hunk 7 above)
Comment (annakhm) [reply website/docs/d/policy_vpc.html.markdown:44]: @tvigneron could you please give your opinion?

**Review by annakhm — APPROVED** [review]
